### PR TITLE
[WIP] Implement reductor task

### PR DESCRIPTION
**Pull request by tfarago**
Here we go, min, max, sum, mean are enough for me for now. I use the metadata because after some thinking I though that would be cleaner than outputting one pixel. And the computed values actually are properties of the data, so, yeah.
I was thinking of implementing also variance but that is a little too much for this task. I would rather create a task *statistics* which can use this task and do their business there. I see something like `ufo-launch dummy-data ! reduce mode=mean ! statistics mode=var ! ...`.
Any remarks?
I know there is the parallel_sum_2d which is very similar, but that's needed by the stitching and I wanted to have this without the need of 2D input data.

### CODE REVIEW

File: src/ufo-reduce-task.c [hunk 1]
```diff
@@ -27,6 +28,28 @@
...
+typedef enum {
+    MODE_SUM,
+    MODE_MEAN,
+    MODE_MIN,
+    MODE_MAX,
+} Mode;
+
+static GEnumValue mode_values[] = {
+    { MODE_SUM,  "MODE_SUM",  "sum" },
+    { MODE_MEAN, "MODE_SUM", "mean" },
```
Comment (tfarago) [inline src/ufo-reduce-task.c]: Will I go straight to hell for this? It makes setup cleaner...

File: src/ufo-reduce-task.c [hunk 2]
```diff
@@ -115,7 +115,7 @@ static guint
 ufo_reduce_task_get_num_dimensions (UfoTask *task,
                                     guint input)
 {
-    return 2;
+    return -1;
```
Comment (tfarago) [inline src/ufo-reduce-task.c]: Is this the correct way? I didn't find it in the docs...

File: src/ufo-reduce-task.c (hunk 1 above)
Comment (matze) [reply src/ufo-reduce-task.c]: No, why? I was doing something similar in the flatten thing AFAIK.

File: src/ufo-reduce-task.c [hunk 3]
```diff
@@ -48,16 +78,31 @@ ufo_reduce_task_setup (UfoTask *task,
...
+    UfoReduceTaskPrivate *priv = UFO_REDUCE_TASK_GET_PRIVATE (task);
+    UfoGpuNode *node = UFO_GPU_NODE (ufo_task_node_get_proc_node (UFO_TASK_NODE (task)));
+    local_size_gvalue = ufo_gpu_node_get_info (node, UFO_GPU_NODE_INFO_MAX_WORK_GROUP_SIZE);
+    priv->local_size = g_value_get_ulong (local_size_gvalue);
+    g_value_unset (local_size_gvalue);
+
+    priv->result = NULL;
+    priv->kernel = ufo_resources_get_kernel (resources, "reductor.cl",
+                                             g_strconcat ("reduce_", mode_values[priv->mode].value_name, NULL),
+                                             error);
```
Comment (matze) [inline src/ufo-reduce-task.c]: That is clever :clap: !

File: src/ufo-reduce-task.c [hunk 4]
```diff
@@ -85,23 +130,154 @@ ufo_reduce_task_process (UfoTask *task,
...
+                                                    0,
+                                                    sizeof (cl_float),
+                                                    &result,
+                                                    0, NULL, NULL));
+    if (priv->mode == MODE_MEAN) {
+        result /= input_size;
     }
 
+    /* Pass the original data on intact */
+    ufo_buffer_copy (inputs[0], output);
```
Comment (matze) [inline src/ufo-reduce-task.c]: Could you check if the recently added `ufo_buffer_swap_data` call works here as well? That would avoid the copy.

**Comment by matze** [comment]
> I was thinking of implementing also variance but that is a little too much for this task. I would rather create a task statistics which can use this task and do their business there.

Please have a look at the contributed filters and see if you could re-use them or base your work on them.

A general question to this PR: do you foresee somehow to reduce in time/across the stream like the `flatten` task?

**Comment by tfarago** [comment]
Tomorrow I will do all what you requested. Now I realized two things which could cause trouble
- in the end I want to use the results in python and I don't want to copy the original image back via the output task. Is it possible to just get the metadata without the actual data? If not, I will fall back to having a UfoBuffer with one pixel.
- with the statistics it's going to be messy because I might use the mean computed here but then still need to do reduce once more with e.g. (data - mean)^2 for variance. Plus the statistics task would require the reductor task in order to have the metadata, which is against the encapsulation principle. So in the end, I think it might make sense to include the statistics here, which will blow up the filter on the one hand, but will be easier for the user on the other.

**Comment by tfarago** [comment]
> Please have a look at the contributed filters and see if you could re-use them or base your work on them.

Did you have the measure task in mind? This would replace that since it's on GPU.

**Comment by tfarago** [comment]
> A general question to this PR: do you foresee somehow to also reduce in time/across the stream

This would of course be nice, I will think about how to do it.

**Comment by tfarago** [comment]
BTW does the tracing of OpenCL code work for you? It did for me yesterday, then I reconfigued cmake after I installed gsl to check the timings of it, I allowed tracing in the configuration but the opencl* trace file is empty.

**Comment by tfarago** [comment]
Alright, new plan. This will stay a reductor thing, i.e. min, max, sum, mean. For the statistics I will update the `ufo-measure-task`. Sounds OK?

**Comment by matze** [comment]
> BTW does the tracing of OpenCL code work for you?

Did you run it on a multi-GPU system? I opened an issue abut that a while ago: https://github.com/ufo-kit/ufo-core/issues/122

> This would of course be nice, I will think about how to do it.

I suppose this will take a while? Or do you want this already merged?

**Comment by tfarago** [comment]
Ahaaa, now it all makes sense. No no, this PR will anyway go away probably.

**Comment by tfarago** [comment]
Superseded by #156

**Pull request closed without merging**